File: models/Order.js
const mongoose = require('mongoose');

const orderSchema = new mongoose.Schema({
  orderNumber: {
    type: String,
    required: true,
    unique: true
  },
  supplier: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Contact',
    required: true
  },
  // Custom supplier information for non-contact suppliers
  customSupplier: {
    name: {
      type: String,
      required: function() { return !this.supplier; }
    },
    company: String,
    contactInfo: {
      email: String,
      phone: String,
      address: {
        street: String,
        city: String,
        state: String,
        zipCode: String,
        country: { type: String, default: 'USA' }
      }
    },
    website: String,
    notes: String
  },
  parts: [{
    part: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Part'
    },
    // Custom part fields for non-catalog items
    customPart: {
      name: {
        type: String,
        required: function() { return !this.part; }
      },
      description: String,
      partNumber: String,
      brand: String,
      category: String,
      image: String, // URL to uploaded image
      specifications: {
        weight: Number,
        dimensions: {
          length: Number,
          width: Number,
          height: Number
        },
        material: String,
        color: String
      }
    },
    quantity: {
      type: Number,
      required: true,
      min: 1
    },
    unitPrice: {
      type: Number,
      required: true,
      min: 0
    },
    totalPrice: {
      type: Number,
      required: true,
      min: 0
    }
  }],
  totalAmount: {
    type: Number,
    required: true,
    min: 0
  },
  status: {
    type: String,
    enum: ['pending', 'confirmed', 'shipped', 'delivered', 'cancelled'],
    default: 'pending'
  },
  progress: {
    type: String,
    enum: ['not_started', 'waiting_on_parts', 'started', 'finished', 'waiting_for_pickup'],
    default: 'not_started'
  },
  orderDate: {
    type: Date,
    default: Date.now
  },
  expectedDeliveryDate: {
    type: Date
  },
  estimatedArrivalTime: {
    type: String, // e.g., "3-5 business days", "1 week", "2-3 weeks"
    required: true
  },
  actualDeliveryDate: {
    type: Date
  },
  shippingInfo: {
    trackingNumber: String,
    carrier: String,
    shippingMethod: String,
    shippingCost: {
      type: Number,
      default: 0
    }
  },
  customTimeLimit: {
    type: Number, // in hours
    default: 72 // 3 days default
  },
  countdownEndTime: {
    type: Date
  },
  notes: String,
  createdBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  }
}, {
  timestamps: true
});

// Pre-save middleware to generate order number and set countdown
orderSchema.pre('save', function(next) {
  if (this.isNew) {
    // Generate order number
    const timestamp = Date.now().toString().slice(-6);
    this.orderNumber = `ORD-${timestamp}`;
    
    // Set countdown end time if custom time limit is provided
    if (this.customTimeLimit) {
      this.countdownEndTime = new Date(Date.now() + (this.customTimeLimit * 60 * 60 * 1000));
    }
  }
  next();
});

// Virtual for countdown status
orderSchema.virtual('countdownStatus').get(function() {
  if (!this.countdownEndTime) return null;
  
  const now = new Date();
  const timeLeft = this.countdownEndTime - now;
  
  if (timeLeft <= 0) {
    return {
      status: 'overdue',
      timeLeft: 0,
      message: 'Package is overdue'
    };
  }
  
  const hours = Math.floor(timeLeft / (1000 * 60 * 60));
  const minutes = Math.floor((timeLeft % (1000 * 60 * 60)) / (1000 * 60));
  
  return {
    status: hours < 24 ? 'urgent' : 'normal',
    timeLeft: timeLeft,
    hours: hours,
    minutes: minutes,
    message: `${hours}h ${minutes}m remaining`
  };
});

// Virtual for days until delivery
orderSchema.virtual('daysUntilDelivery').get(function() {
  if (!this.expectedDeliveryDate) return null;
  
  const now = new Date();
  const timeDiff = this.expectedDeliveryDate - now;
  return Math.ceil(timeDiff / (1000 * 60 * 60 * 24));
});

module.exports = mongoose.model('Order', orderSchema);

File: flask-app/models/order.py
from database import db
from datetime import datetime, timedelta
from sqlalchemy import Index
import random
import string

class Order(db.Model):
    __tablename__ = 'orders'
    
    id = db.Column(db.Integer, primary_key=True)
    order_number = db.Column(db.String(50), unique=True, nullable=False, index=True)
    
    # Supplier Information
    supplier_id = db.Column(db.Integer, db.ForeignKey('contacts.id'))
    
    # Custom supplier information for non-contact suppliers (stored as JSON)
    custom_supplier = db.Column(db.JSON)
    
    # Parts information (stored as JSON array)
    parts = db.Column(db.JSON, nullable=False)  # Array of part objects
    
    total_amount = db.Column(db.Float, nullable=False)
    
    status = db.Column(db.Enum('pending', 'confirmed', 'shipped', 'delivered', 'cancelled', 
                              name='order_statuses'), default='pending')
    progress = db.Column(db.Enum('not_started', 'waiting_on_parts', 'started', 'finished', 
                                'waiting_for_pickup', name='order_progress'), default='not_started')
    
    # Dates
    order_date = db.Column(db.DateTime, default=datetime.utcnow)
    expected_delivery_date = db.Column(db.DateTime)
    estimated_arrival_time = db.Column(db.String(100), nullable=False)  # e.g., "3-5 business days"
    actual_delivery_date = db.Column(db.DateTime)
    
    # Shipping Information (stored as JSON)
    shipping_info = db.Column(db.JSON)  # tracking_number, carrier, shipping_method, shipping_cost
    
    # Countdown Information
    custom_time_limit = db.Column(db.Integer, default=72)  # in hours
    countdown_end_time = db.Column(db.DateTime)
    
    notes = db.Column(db.Text)
    
    # Foreign key to creator
    created_by_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
    
    # Timestamps
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    def __init__(self, **kwargs):
        super(Order, self).__init__(**kwargs)
        if not self.order_number:
            self.generate_order_number()
        if self.custom_time_limit and not self.countdown_end_time:
            self.countdown_end_time = datetime.utcnow() + timedelta(hours=self.custom_time_limit)
    
    def generate_order_number(self):
        """Generate unique order number"""
        timestamp = str(int(datetime.utcnow().timestamp()))[-6:]
        self.order_number = f"ORD-{timestamp}"
    
    @property
    def countdown_status(self):
        """Get countdown status information"""
        if not self.countdown_end_time:
            return None
        
        now = datetime.utcnow()
        time_left = self.countdown_end_time - now
        
        if time_left.total_seconds() <= 0:
            return {
                'status': 'overdue',
                'time_left': 0,
                'message': 'Package is overdue'
            }
        
        total_seconds = int(time_left.total_seconds())
        hours = total_seconds // 3600
        minutes = (total_seconds % 3600) // 60
        
        return {
            'status': 'urgent' if hours < 24 else 'normal',
            'time_left': total_seconds,
            'hours': hours,
            'minutes': minutes,
            'message': f"{hours}h {minutes}m remaining"
        }
    
    @property
    def days_until_delivery(self):
        """Calculate days until expected delivery"""
        if not self.expected_delivery_date:
            return None
        
        now = datetime.utcnow()
        time_diff = self.expected_delivery_date - now
        return max(0, time_diff.days)
    
    def to_dict(self):
        """Convert order object to dictionary"""
        return {
            'id': self.id,
            'order_number': self.order_number,
            'supplier_id': self.supplier_id,
            'custom_supplier': self.custom_supplier,
            'parts': self.parts or [],
            'total_amount': self.total_amount,
            'status': self.status,
            'progress': self.progress,
            'order_date': self.order_date.isoformat() if self.order_date else None,
            'expected_delivery_date': self.expected_delivery_date.isoformat() if self.expected_delivery_date else None,
            'estimated_arrival_time': self.estimated_arrival_time,
            'actual_delivery_date': self.actual_delivery_date.isoformat() if self.actual_delivery_date else None,
            'shipping_info': self.shipping_info or {},
            'custom_time_limit': self.custom_time_limit,
            'countdown_end_time': self.countdown_end_time.isoformat() if self.countdown_end_time else None,
            'countdown_status': self.countdown_status,
            'days_until_delivery': self.days_until_delivery,
            'notes': self.notes,
            'created_by_id': self.created_by_id,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None
        }
    
    def __repr__(self):
        return f'<Order {self.order_number}>'

# Create indexes for better query performance
Index('idx_order_number', Order.order_number)
Index('idx_order_status', Order.status)
Index('idx_order_date', Order.order_date) 